File: csharp/src/Vista.SDK/Experimental/Transport/DataList/DataList.cs
namespace Vista.SDK.Experimental.Transport.DataList;

public sealed record DataListPackage(Package Package);

public sealed record Package(Header Header, DataList DataList);

public sealed record Header(
    AssetIdentifier AssetId,
    ConfigurationReference DataListId,
    VersionInformation? VersionInformation,
    string? Author,
    DateTimeOffset? DateCreated,
    IReadOnlyDictionary<string, object> CustomHeaders
);

public sealed record ConfigurationReference(string Id, string? Version, DateTimeOffset TimeStamp);

public sealed record VersionInformation(string NamingRule, string NamingSchemeVersion, string? ReferenceUrl);

public sealed record DataList(IReadOnlyList<Data> Data);

public sealed record Data(DataId DataId, Property Property);

public sealed record DataId(ILocalId LocalId, string? ShortId, NameObject? NameObject);

public sealed record NameObject(string NamingRule, IReadOnlyDictionary<string, object> CustomProperties);

public sealed record DataType(string Type, double? UpdateCycle, double? CalculationPeriod);

public sealed record Restriction(
    IReadOnlyList<string>? Enumeration,
    int? FractionDigits,
    int? Length,
    double? MaxExclusive,
    double? MaxInclusive,
    int? MaxLength,
    double? MinExclusive,
    double? MinInclusive,
    int? MinLength,
    string? Pattern,
    int? TotalDigits,
    WhiteSpace? WhiteSpace
);

public enum WhiteSpace
{
    Preserve = 0,
    Replace = 1,
    Collapse = 2,
}

public sealed record Format(string Type, Restriction? Restriction);

public sealed record Range(double High, double Low);

public sealed record Unit(
    string UnitSymbol,
    string? QuantityName,
    IReadOnlyDictionary<string, object> CustomProperties
);

public sealed record Property(
    DataType DataType,
    Format Format,
    Range? Range,
    Unit? Unit,
    string? QualityCoding,
    string? AlertPriority,
    string? Name,
    string? Remarks,
    IReadOnlyDictionary<string, object> CustomProperties
);

File: js/lib/experimental/transport/domain/data-list/DataList.ts
import { ILocalId } from "../../../../ILocalId";
import { Version } from "../../../../transport/domain/data-channel/Version";
import { AssetIdentifier } from "../AssetIdentifier";

export namespace DataList {
    export type DataListPackage = {
        package: Package;
    };

    export type Package = {
        header: Header;
        dataList: DataList;
    };

    export type Header = {
        assetId: AssetIdentifier;
        dataListId: ConfigurationReference;
        versionInformation?: VersionInformation;
        author?: string;
        dateCreated?: Date;
        customHeaders?: { [k: string]: unknown };
    };

    export type ConfigurationReference = {
        id: string;
        version?: Version;
        timestamp: Date;
    };

    export type VersionInformation = {
        namingRule: string;
        namingSchemeVersion: string;
        referenceUrl?: string;
    };

    export type DataList = {
        data: Data[];
    };

    export type Data = {
        dataId: DataId;
        property: Property;
    };

    export type DataId = {
        localId: ILocalId;
        shortId?: string;
        nameObject?: NameObject;
    };

    export type NameObject = {
        namingRule: string;
        customProperties?: { [k: string]: unknown };
    };

    export type Property = {
        dataType: DataType;
        format: Format;
        range?: Range;
        unit?: Unit;
        qualityCoding?: string;
        alertPriority?: string;
        name?: string;
        remarks?: string;
        customProperties?: { [k: string]: unknown };
    };

    export type Format = {
        type: string;
        restriction?: Restriction;
    };

    export type Range = {
        high: string;
        low: string;
    };

    export type Unit = {
        unitSymbol: string;
        quantityName?: string;
        customProperties?: { [k: string]: unknown };
    };

    export type DataType = {
        type: string;
        updateCycle?: string;
        calculationPeriod?: string;
    };

    export type Restriction = {
        enumeration?: string[];
        fractionDigits?: string;
        length?: string;
        maxExclusive?: string;
        maxInclusive?: string;
        maxLength?: string;
        minExclusive?: string;
        minInclusive?: string;
        minLength?: string;
        pattern?: string;
        totalDigits?: string;
        whiteSpace?: "Preserve" | "Replace" | "Collapse";
    };
}
